### PR TITLE
Static side effect analysis for ASL

### PR DESCRIPTION
**Pull request by HadrienRenaud**
Todo:

- [x] Configs
- [x] Non determinism
- [x] Check all possible constructions

### CODE REVIEW

**Review by Roman-Manevich — COMMENTED** [review]
Some comments for now

**Review by Roman-Manevich — COMMENTED** [review]
A documentation-related comment

File: asllib/SideEffect.ml [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+    | Constant, Constant | Config, Config -> true
+    | Execution b1, Execution b2 -> Bool.equal b1 b2
+    | Constant, (Config | Execution _)
+    | Config, (Constant | Execution _)
+    | Execution _, (Config | Constant) ->
+        false
+
+  let is_before t1 t2 =
+    match (t1, t2) with
+    | Constant, Constant | Config, Config | Execution _, Execution _ -> true
```
Comment (Roman-Manevich) [inline asllib/SideEffect.ml:20]: So both `Execution(true) < Execution(false)` and `Execution(false) < Execution(true)`, which means this is not a total order. It's a little strange.

**Review by Roman-Manevich — COMMENTED** [review]
Another comment

**Review by hrutvik — COMMENTED** [review]
Mostly minor comments, and a few queries!

File: asllib/Typing.ml [hunk 2]
```diff
@@ -542,41 +565,42 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
...
-  let is_statically_evaluable ~loc env e =
-    let use_set = use_e e ISet.empty in
-    ISet.for_all (storage_is_pure ~loc env) use_set
-    |: TypingRule.IsStaticallyEvaluable
-  (* End *)
-
   (* Begin CheckStaticallyEvaluable *)
-  let check_statically_evaluable (env : env) e () =
-    if is_statically_evaluable ~loc:e env e then
+  let check_statically_evaluable e ses () =
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:569]: Can we keep the convention that annotated types used just for errror report be named something like `loc` or `error_annot`?

File: asllib/Typing.ml [hunk 3]
```diff
@@ -3297,18 +3568,19 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
...
                   T_Exception (fields @ extra_fields) |> add_pos_from_st ty
               | Some _ -> conflict loc [ T_Record []; T_Exception [] ] ty
               | None -> undefined_identifier loc super
           and env = add_subtype name super env in
           (env, new_ty)
       (* AnnotateExtraFields) *)
     in
-    let t2 = annotate_type ~decl:true ~loc env1 t1 in
-    let env2 = add_type name t2 env1 in
+    let t2, ses_t = annotate_type ~decl:true ~loc env1 t1 in
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:3581]: Don't we need to check that type definitions side effects are statically evaluable?

File: asllib/Typing.ml [hunk 4]
```diff
@@ -2217,18 +2336,24 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
                 (Error.BadArity
                    (Static, "LEDestructuring", List.length tys, List.length les))
             else
-              let les' = List.map2 (annotate_lexpr env) les tys in
-              LE_Destructuring les' |> here
+              let les', sess =
+                List.map2 (annotate_lexpr env) les tys |> List.split
+              in
+              let ses =
+                (* TODO left-hand-side conflicting union *)
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:2343]: I don't understand the comment. What's precenting you from using conflicting union here?

File: asllib/Typing.ml [hunk 5]
```diff
@@ -2237,24 +2362,28 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
...
+              best_effort (slices, SES.empty) @@ fun _ ->
+              annotate_slices env slices ~loc
+            in
             let+ () = check_disjoint_slices le env slices2 in
             let+ () =
               check_true (not (list_is_empty slices)) @@ fun () ->
               fatal_from le Error.EmptySlice
             in
-            LE_Slice (le2, slices2) |> here |: TypingRule.LESlice
+            let ses = SES.union ses1 ses2 in
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:2374]: I think the semantics evaluates the base of the slice and the slices in parallel so probably you should use `conflicting_union`, no?

File: asllib/Typing.ml [hunk 6]
```diff
@@ -2357,13 +2488,19 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
     if can_be_initialized_with env s t then () else conflict loc [ s.desc ] t
   (* End *)
 
+  let should_remember_immutable_expression ses =
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:2491]: how about `statically_evaluable_modulo_assertions`?

File: asllib/Typing.ml [hunk 7]
```diff
@@ -2357,13 +2488,19 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
...
+    SES.is_statically_evaluable ses_non_assert
+
   (* Begin AddImmutableExpr *)
-  let add_immutable_expr ~loc env ldk e_opt x =
-    match (ldk, e_opt) with
-    | (LDK_Constant | LDK_Let), Some e when is_statically_evaluable ~loc env e
-      ->
-        let e' = StaticModel.try_normalize env e in
-        add_local_immutable_expr x e' env |: TypingRule.AddImmutableExpr
+  let add_immutable_expressions env ldk typed_e_opt x =
```
Comment (Roman-Manevich) [inline asllib/Typing.ml:2496]: why does the name refer to the plural expressions if there's just one?

**Review by Roman-Manevich — APPROVED** [review]
Looks good to me!

**Pull request merged**